### DISCUSSION TITLE
pipe rg filenames into to itself

### DISCUSSION THREAD

**Discussion by chuanqisun** [BurntSushi/ripgrep#1772] [Q&A] [answered]
Dear ripgrep maintainers,

I currently use multiple ripgrep executions to emulate the "AND" logic operator, like this:

`rg foo -l --null-data | xargs -0 rg bar -l --null-data | xargs -0 rg magic ...`

This technique works well for me but I was hoping to remove the dependency on `xargs` so I can ship my code to windows and linux without adding platform specific utils.

I am using `-l` because I hope to search the entire file, not just the previously matched lines.

My question is 

1. Is there a way to pipe rg's output into itself without additional dependencies?
2. If not, would you be willing to extend `rg` so we can get an output that is compatible with the expected input format?

I feel `-0`, `--null-data`, and `--crlf` already manipulate line terminators, so it could be a small change.

Thanks for the amazing work.

**Comment by chuanqisun** [comment]
I'm sorry I just realized a bigger problem is the positioning of the argument, instead of the line terminators. When using pipe, `rg` cannot treat the input as filenames unless I manipulate it with `xargs`. I'm curious if there is any workaround for this. Again I'm sorry I have understated the problem here.

**Comment by BurntSushi** [comment, marked as answer]
Thanks for the question.

There is an issue tracking boolean matching that would permit one to express things like "matches foo and bar," but that's intended to work at a line level. Working at the file level as you describe here is a different endeavor, and no, there are no plans to add support for this use case to ripgrep. Using xargs with a pipeline is the best solution. But if you need to do this on Windows, you can, it will just be more verbose.

I don't understand your suggestion about line terminators. I don't see how that would resolve your problem. Please provide an example of what would solve your problem.

As a small side note, you can solve a subset of cases via a multiline regex. For example, to match both foo and bar, you would use `(?s)foo.*bar|bar.*foo`. But the size of the regex obviously grows quite quickly as you add more patterns.

**Reply by chuanqisun** [reply]
Thanks for the pro tip on the New Years Eve! Please dismiss the line terminator question. I didn't understand xargs when I asked it.  I should have read the man page more carefully.

Cheers!
